src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/WorkspaceAwareDecoratingFeatureSource.java

```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Override
+    public SimpleFeatureCollection getFeatures(Query query) throws IOException {
+        query = appendWorkspaceToQuery(query);
+        return delegate.getFeatures(query);
+    }
+
+    private Set<FeatureId> getCollectionIdsForWorkspace() throws IOException {
+        String workspace = null;
+        Set<FeatureId> ids = new LinkedHashSet<>();
+        if (workspaceInfo != null) {
+            SimpleFeatureCollection idFeatureCollection = getIdFeatureCollection();
+            idFeatureCollection.accepts(f -> ids.add(f.getIdentifier()), null);
+        }
+        return ids;
+    }
+
+    private Set<String> getCollectionNamesForWorkspace() throws IOException {
+        String workspace = null;
+        Set<String> names = new LinkedHashSet<>();
+        if (workspaceInfo != null) {
+            SimpleFeatureCollection idFeatureCollection = getIdFeatureCollection();
+            idFeatureCollection.accepts(
+                    f -> names.add(f.getProperty("name").getValue().toString()), null);
+        }
+        return names;
+    }
+
+    private SimpleFeatureCollection getIdFeatureCollection() throws IOException {
+        String workspace = workspaceInfo.getName();
+        Query idsQuery = new Query();
+        idsQuery.setFilter(
+                FF.or(
+                        FF.or(
+                                FF.isNull(FF.property(WORKSPACES_FIELD)),
+                                FF.equals(
+                                        FF.function("arrayhasnull", FF.property(WORKSPACES_FIELD)),
+                                        FF.literal(true))),
+                        FF.equals(FF.property(WORKSPACES_FIELD), FF.literal(workspace))));
+        SimpleFeatureSource collectionSource =
+                openSearchAccess
+                        .getDelegateStore()
+                        .getFeatureSource(JDBCOpenSearchAccess.COLLECTION);
+        return collectionSource.getFeatures(idsQuery);
+    }
+
+    private Filter appendWorkspaceToFilter(Filter filterIn) throws IOException {
+        if (delegate.getSchema().getTypeName().equals(JDBCOpenSearchAccess.COLLECTION)) {
+            Set<FeatureId> collectionIds = getCollectionIdsForWorkspace();
+            if (collectionIds != null && !collectionIds.isEmpty()) {
+                return appendWorkspaceIdsToFilter(filterIn, collectionIds);
+            }
+        } else if (delegate.getSchema().getTypeName().equals(JDBCOpenSearchAccess.PRODUCT)) {
+            Set<String> collectionNames = getCollectionNamesForWorkspace();
+            if (collectionNames != null && !collectionNames.isEmpty()) {
+                return appendWorkspaceNamesToFilter(filterIn, collectionNames);
+            }
+        }
+        return filterIn;
+    }
+
+    private Filter appendWorkspaceIdsToFilter(Filter filterIn, Set<FeatureId> collectionIds) {
+        if (collectionIds == null || collectionIds.isEmpty()) {
```

Comment:
Hum... this is an intesting bit... if method loading collections for the current workspace did not find anything, it means there was no collection either with a null, or with a null entry, or with a matching entry.... hence, shouldn't the response be completely empty instead?

Reply:
Thinking out loud.. perhaps it's easier to have a builder method taking the workspace info, loading the authorized collections, and if it's all empty, it just returns ``DataUtilities.source(new EmptyFeatureCollection(schema))``, otherwise it creates this wrapper with the list of collection features.

As an alternative, make this method return Filter.EXCLUDE.